<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper plugin-docs plugin-id-default docs-version-current docs-doc-page docs-doc-id-tucca-rna-seq/enrichment-analysis" data-has-hydrated="false">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v3.8.1">
<title data-rh="true">Functional Enrichment Analysis | TUCCA Bioinformatics Docs</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:image" content="https://tucca-cellag.github.io/img/Center for Cellular Agriculture_blk+bl_vert.png"><meta data-rh="true" name="twitter:image" content="https://tucca-cellag.github.io/img/Center for Cellular Agriculture_blk+bl_vert.png"><meta data-rh="true" property="og:url" content="https://tucca-cellag.github.io/tucca-rna-seq/enrichment-analysis"><meta data-rh="true" property="og:locale" content="en"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Functional Enrichment Analysis | TUCCA Bioinformatics Docs"><meta data-rh="true" name="description" content="This documentation page is currently under active development. The information"><meta data-rh="true" property="og:description" content="This documentation page is currently under active development. The information"><link data-rh="true" rel="canonical" href="https://tucca-cellag.github.io/tucca-rna-seq/enrichment-analysis"><link data-rh="true" rel="alternate" href="https://tucca-cellag.github.io/tucca-rna-seq/enrichment-analysis" hreflang="en"><link data-rh="true" rel="alternate" href="https://tucca-cellag.github.io/tucca-rna-seq/enrichment-analysis" hreflang="x-default"><script data-rh="true" type="application/ld+json">{"@context":"https://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Functional Enrichment Analysis","item":"https://tucca-cellag.github.io/tucca-rna-seq/enrichment-analysis"}]}</script><link rel="stylesheet" href="/assets/css/styles.2d474d0a.css">
<script src="/assets/js/runtime~main.23394ea9.js" defer="defer"></script>
<script src="/assets/js/main.29038db3.js" defer="defer"></script>
</head>
<body class="navigation-with-keyboard">
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;"><defs>
<symbol id="theme-svg-external-link" viewBox="0 0 24 24"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"/></symbol>
</defs></svg>
<script>!function(){var t="light";var e=function(){try{return new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}}()||function(){try{return window.localStorage.getItem("theme")}catch(t){}}();document.documentElement.setAttribute("data-theme",e||t),document.documentElement.setAttribute("data-theme-choice",e||t)}(),function(){try{const c=new URLSearchParams(window.location.search).entries();for(var[t,e]of c)if(t.startsWith("docusaurus-data-")){var a=t.replace("docusaurus-data-","data-");document.documentElement.setAttribute(a,e)}}catch(t){}}()</script><div id="__docusaurus"><div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="theme-layout-navbar navbar navbar--fixed-top"><div class="navbar__inner"><div class="theme-layout-navbar-left navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/"><b class="navbar__title text--truncate">TUCCA Bioinformatics Docs</b></a><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/tucca-rna-seq/introduction">tucca-rna-seq</a><a class="navbar__item navbar__link" href="/helpful-resources/learn-about-cell-ag">Helpful Resources</a><a class="navbar__item navbar__link" href="/publications/ast1">Publications</a><a class="navbar__item navbar__link" href="/our-team">Our Team</a></div><div class="theme-layout-navbar-right navbar__items navbar__items--right"><a href="https://github.com/tucca-cellag" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">Our GitHub Organization<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="system mode" aria-label="Switch between dark and light mode (currently system mode)"><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP systemToggleIcon_QzmC"><path fill="currentColor" d="m12 21c4.971 0 9-4.029 9-9s-4.029-9-9-9-9 4.029-9 9 4.029 9 9 9zm4.95-13.95c1.313 1.313 2.05 3.093 2.05 4.95s-0.738 3.637-2.05 4.95c-1.313 1.313-3.093 2.05-4.95 2.05v-14c1.857 0 3.637 0.737 4.95 2.05z"></path></svg></button></div><div class="navbarSearchContainer_Bca1"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="theme-layout-main main-wrapper mainWrapper_z2l0"><div class="docsWrapper_hBAB"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docRoot_UBD9"><aside class="theme-doc-sidebar-container docSidebarContainer_YfHR"><div class="sidebarViewport_aRkj"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/tucca-rna-seq/introduction">Introduction</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a href="https://github.com/tucca-cellag/tucca-rna-seq" target="_blank" rel="noopener noreferrer" class="menu__link menuExternalLink_NmtK">GitHub Repository<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" role="button" aria-expanded="false" href="/tucca-rna-seq/data-collection/data-collection-why">Data Collection 101 for RNA-Seq</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" role="button" aria-expanded="false" href="/tucca-rna-seq/install-and-config/snakemake-primer">Installation and Configuration of tucca-rna-seq</a></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/tucca-rna-seq/running">Running tucca-rna-seq</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" href="/tucca-rna-seq/enrichment-analysis">Functional Enrichment Analysis</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/tucca-rna-seq/tufts-specific/hpc-quick-start">For Tufts Users</a><button aria-label="Expand sidebar category &#x27;For Tufts Users&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/tucca-rna-seq/citing-the-workflow">Citing the Workflow</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/tucca-rna-seq/help">Getting Help</a></li></ul></nav></div></div></aside><main class="docMainContainer_TBSr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link">Functional Enrichment Analysis</span></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><div class="theme-admonition theme-admonition-warning admonition_xJq3 alert alert--warning"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 16 16"><path fill-rule="evenodd" d="M8.893 1.5c-.183-.31-.52-.5-.887-.5s-.703.19-.886.5L.138 13.499a.98.98 0 0 0 0 1.001c.193.31.53.501.886.501h13.964c.367 0 .704-.19.877-.5a1.03 1.03 0 0 0 .01-1.002L8.893 1.5zm.133 11.497H6.987v-2.003h2.039v2.003zm0-3.004H6.987V5.987h2.039v4.006z"></path></svg></span>Page in Progress</div><div class="admonitionContent_BuS1"><p>This documentation page is currently under active development. The information
may be incomplete or subject to change. Thank you for your patience as we work
to improve our guides.</p></div></div>
<header><h1>Functional Enrichment Analysis</h1></header>
<p>The <code>tucca-rna-seq</code> workflow provides comprehensive functional enrichment
analysis to help interpret the biological meaning of differentially expressed
genes. This guide covers all enrichment analysis capabilities and how to
configure and interpret results.</p>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="overview">Overview<a href="#overview" class="hash-link" aria-label="Direct link to Overview" title="Direct link to Overview">​</a></h2>
<p>Functional enrichment analysis identifies biological processes, pathways, and
gene sets that are over-represented in your differentially expressed genes
compared to a background set. This helps you understand:</p>
<ul>
<li><strong>Biological significance</strong> of your gene expression changes</li>
<li><strong>Pathway involvement</strong> in your experimental conditions</li>
<li><strong>Gene set associations</strong> with specific biological functions</li>
<li><strong>Tissue-specific patterns</strong> in gene expression</li>
</ul>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="analysis-types">Analysis Types<a href="#analysis-types" class="hash-link" aria-label="Direct link to Analysis Types" title="Direct link to Analysis Types">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="1-standard-enrichment-analysis-clusterprofiler">1. Standard Enrichment Analysis (clusterProfiler)<a href="#1-standard-enrichment-analysis-clusterprofiler" class="hash-link" aria-label="Direct link to 1. Standard Enrichment Analysis (clusterProfiler)" title="Direct link to 1. Standard Enrichment Analysis (clusterProfiler)">​</a></h3>
<p>The workflow uses <code>clusterProfiler</code> for comprehensive enrichment analysis:</p>
<div class="theme-admonition theme-admonition-tip admonition_xJq3 alert alert--success"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>See it in Action</div><div class="admonitionContent_BuS1"><p>The Over-Representation Analysis (ORA) and Gene Set Enrichment Analysis (GSEA)
features of this workflow were critical in uncovering the molecular mechanisms
of cellular adaptation in a recent study. See the <a href="/publications/ast1">publication case study</a>
for details on how these tools led to novel biological insights.</p></div></div>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="gene-ontology-go-analysis">Gene Ontology (GO) Analysis<a href="#gene-ontology-go-analysis" class="hash-link" aria-label="Direct link to Gene Ontology (GO) Analysis" title="Direct link to Gene Ontology (GO) Analysis">​</a></h4>
<ul>
<li><strong>Biological Process (BP)</strong>: Cellular processes and biological functions</li>
<li><strong>Molecular Function (MF)</strong>: Biochemical activities and molecular interactions</li>
<li><strong>Cellular Component (CC)</strong>: Subcellular locations and cellular structures</li>
</ul>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="kegg-pathway-analysis">KEGG Pathway Analysis<a href="#kegg-pathway-analysis" class="hash-link" aria-label="Direct link to KEGG Pathway Analysis" title="Direct link to KEGG Pathway Analysis">​</a></h4>
<ul>
<li><strong>Metabolic pathways</strong>: Energy metabolism, biosynthesis</li>
<li><strong>Signaling pathways</strong>: Cell communication, signal transduction</li>
<li><strong>Disease pathways</strong>: Disease-associated gene sets</li>
</ul>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="wikipathways-analysis">WikiPathways Analysis<a href="#wikipathways-analysis" class="hash-link" aria-label="Direct link to WikiPathways Analysis" title="Direct link to WikiPathways Analysis">​</a></h4>
<ul>
<li><strong>Community-curated pathways</strong>: Expert-validated pathway databases</li>
<li><strong>Species-specific pathways</strong>: Tailored to your organism of interest</li>
<li><strong>Dynamic pathway updates</strong>: Regular community updates</li>
</ul>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="kegg-module-analysis">KEGG Module Analysis<a href="#kegg-module-analysis" class="hash-link" aria-label="Direct link to KEGG Module Analysis" title="Direct link to KEGG Module Analysis">​</a></h4>
<ul>
<li><strong>Functional modules</strong>: Subsets of KEGG pathways</li>
<li><strong>Metabolic networks</strong>: Interconnected metabolic reactions</li>
<li><strong>Regulatory modules</strong>: Gene regulatory networks</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="2-msigdb-molecular-signatures-database">2. MSigDB (Molecular Signatures Database)<a href="#2-msigdb-molecular-signatures-database" class="hash-link" aria-label="Direct link to 2. MSigDB (Molecular Signatures Database)" title="Direct link to 2. MSigDB (Molecular Signatures Database)">​</a></h3>
<p>Access to 8 major gene set collections with over 25,000 gene sets:</p>
<div class="container center"><div class="row"><div class="col"><div class="card item shadow--md"><div class="card__header"><h4>H: Hallmark Gene Sets</h4></div><div class="card__body"><p>50 well-defined gene sets representing specific biological states and processes.</p><ul><li>Apoptosis</li><li>DNA Repair</li><li>Inflammatory Response</li><li>Oxidative Phosphorylation</li></ul></div></div></div><div class="col"><div class="card item shadow--md"><div class="card__header"><h4>C2: Curated Gene Sets</h4></div><div class="card__body"><p>5,922 gene sets from various sources including pathway databases, literature, and expert knowledge.</p><ul><li>KEGG pathways</li><li>Reactome pathways</li><li>Biocarta pathways</li><li>Literature-derived sets</li></ul></div></div></div><div class="col"><div class="card item shadow--md"><div class="card__header"><h4>C5: Ontology Gene Sets</h4></div><div class="card__body"><p>10,922 gene sets derived from Gene Ontology terms.</p><ul><li>Biological processes</li><li>Molecular functions</li><li>Cellular components</li><li>Hierarchical organization</li></ul></div></div></div></div></div>
<p><strong>Additional Collections:</strong></p>
<ul>
<li><strong>C1</strong>: Positional gene sets (299 sets)</li>
<li><strong>C3</strong>: Regulatory target gene sets (3,738 sets)</li>
<li><strong>C4</strong>: Computational gene sets (858 sets)</li>
<li><strong>C6</strong>: Oncogenic signatures (189 sets)</li>
<li><strong>C7</strong>: Immunologic signatures (4,872 sets)</li>
<li><strong>C8</strong>: Cell type signature gene sets (746 sets)</li>
</ul>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="custom-gmt-files">Custom GMT Files<a href="#custom-gmt-files" class="hash-link" aria-label="Direct link to Custom GMT Files" title="Direct link to Custom GMT Files">​</a></h4>
<p>Add your own gene sets in GMT format:</p>
<div class="language-gmt codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-gmt codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">SET_NAME    DESCRIPTION    GENE1    GENE2    GENE3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">MyPathway   Custom pathway ACTB     GAPDH    PAX3</span><br></span></code></pre></div></div>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="3-spia-signaling-pathway-impact-analysis">3. SPIA (Signaling Pathway Impact Analysis)<a href="#3-spia-signaling-pathway-impact-analysis" class="hash-link" aria-label="Direct link to 3. SPIA (Signaling Pathway Impact Analysis)" title="Direct link to 3. SPIA (Signaling Pathway Impact Analysis)">​</a></h3>
<p>Topology-based pathway analysis that considers:</p>
<ul>
<li><strong>Gene interactions</strong> within pathways</li>
<li><strong>Pathway structure</strong> and connectivity</li>
<li><strong>Impact scores</strong> based on topology</li>
<li><strong>Statistical significance</strong> of pathway involvement</li>
</ul>
<div class="theme-admonition theme-admonition-warning admonition_xJq3 alert alert--warning"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 16 16"><path fill-rule="evenodd" d="M8.893 1.5c-.183-.31-.52-.5-.887-.5s-.703.19-.886.5L.138 13.499a.98.98 0 0 0 0 1.001c.193.31.53.501.886.501h13.964c.367 0 .704-.19.877-.5a1.03 1.03 0 0 0 .01-1.002L8.893 1.5zm.133 11.497H6.987v-2.003h2.039v2.003zm0-3.004H6.987V5.987h2.039v4.006z"></path></svg></span>KEGG API Usage Restrictions</div><div class="admonitionContent_BuS1"><p>SPIA uses the KEGG REST API (rest.kegg.jp) which is made available only for
academic use by academic users. Non-academic users must follow the KEGG
Non-academic use guidelines. See <a href="https://www.kegg.jp/kegg/legal.html" target="_blank" rel="noopener noreferrer">KEGG Legal Information</a>
for details.</p></div></div>
<div class="theme-admonition theme-admonition-info admonition_xJq3 alert alert--info"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 14 16"><path fill-rule="evenodd" d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg></span>SPIA Two-Step Process</div><div class="admonitionContent_BuS1"><p>The workflow implements SPIA in two steps:</p><ol>
<li><strong>Data Preparation</strong> (<code>make_spia_data.R</code>): Downloads KEGG pathway XML files
and generates SPIA data files</li>
<li><strong>Analysis</strong> (<code>run_spia.R</code>): Performs SPIA analysis using the pre-generated
data files</li>
</ol></div></div>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="4-harmonizome-database">4. Harmonizome Database<a href="#4-harmonizome-database" class="hash-link" aria-label="Direct link to 4. Harmonizome Database" title="Direct link to 4. Harmonizome Database">​</a></h3>
<p>Tissue-specific gene sets from various expression datasets:</p>
<ul>
<li><strong>GTEx Tissue Gene Expression Profiles</strong>: Human tissue-specific expression</li>
<li><strong>BioGPS Human Cell Type Profiles</strong>: Cell type-specific gene sets</li>
<li><strong>Human Protein Atlas</strong>: Protein expression across tissues</li>
<li><strong>Expression Atlas</strong>: Multi-species expression data</li>
</ul>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="configuration">Configuration<a href="#configuration" class="hash-link" aria-label="Direct link to Configuration" title="Direct link to Configuration">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="basic-configuration">Basic Configuration<a href="#basic-configuration" class="hash-link" aria-label="Direct link to Basic Configuration" title="Direct link to Basic Configuration">​</a></h3>
<p>Configure enrichment analysis in your <code>config.yaml</code>:</p>
<div class="language-yaml codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-yaml codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token key atrule" style="color:#00a4db">enrichment</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><span class="token key atrule" style="color:#00a4db">padj_cutoff</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token number" style="color:#36acaa">0.05</span><span class="token plain">       </span><span class="token comment" style="color:#999988;font-style:italic"># Adjusted p-value cutoff for significant genes</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><span class="token key atrule" style="color:#00a4db">clusterprofiler</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">gsea</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">gseGO</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05, minGSSize = 10, maxGSSize = 500&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">gseKEGG</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05, minGSSize = 10, maxGSSize = 500&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">ora</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">enrichGO</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">enrichKEGG</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">wikipathways</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">enabled</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token boolean important" style="color:#36acaa">true</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">enrichWP</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">gseWP</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05, minGSSize = 10, maxGSSize = 500&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">kegg_module</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">enabled</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token boolean important" style="color:#36acaa">true</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">enrichMKEGG</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">gseMKEGG</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05, minGSSize = 10, maxGSSize = 500&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><span class="token key atrule" style="color:#00a4db">msigdb</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">enabled</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token boolean important" style="color:#36acaa">true</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">collections</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">[</span><span class="token string" style="color:#e3116c">&quot;H&quot;</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;C2&quot;</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;C5&quot;</span><span class="token punctuation" style="color:#393A34">]</span><span class="token plain">     </span><span class="token comment" style="color:#999988;font-style:italic"># Hallmark, Curated, Ontology</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">custom_gmt_files</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">[</span><span class="token punctuation" style="color:#393A34">]</span><span class="token plain">                </span><span class="token comment" style="color:#999988;font-style:italic"># Add custom GMT files here</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">ora</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05, minGSSize = 10, maxGSSize = 500&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">gsea</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05, minGSSize = 10, maxGSSize = 500&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><span class="token key atrule" style="color:#00a4db">spia</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">enabled</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token boolean important" style="color:#36acaa">true</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;beta = NULL, verbose = TRUE, plots = FALSE&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token comment" style="color:#999988;font-style:italic"># When enabled, the workflow automatically:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token comment" style="color:#999988;font-style:italic"># 1. Downloads KEGG pathway data for your organism</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token comment" style="color:#999988;font-style:italic"># 2. Generates SPIA data files using makeSPIAdata()</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token comment" style="color:#999988;font-style:italic"># 3. Performs SPIA analysis using the generated data</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><span class="token key atrule" style="color:#00a4db">harmonizome</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">enabled</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token boolean important" style="color:#36acaa">false</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">datasets</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token punctuation" style="color:#393A34">-</span><span class="token plain"> </span><span class="token key atrule" style="color:#00a4db">name</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;GTEx Tissue Gene Expression Profiles&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        </span><span class="token key atrule" style="color:#00a4db">gene_sets</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">[</span><span class="token string" style="color:#e3116c">&quot;Muscle&quot;</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;Adipose&quot;</span><span class="token punctuation" style="color:#393A34">]</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token punctuation" style="color:#393A34">-</span><span class="token plain"> </span><span class="token key atrule" style="color:#00a4db">name</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;BioGPS Human Cell Type and Tissue Gene Expression Profiles&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        </span><span class="token key atrule" style="color:#00a4db">gene_sets</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">[</span><span class="token string" style="color:#e3116c">&quot;Adipocyte&quot;</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;Myocyte&quot;</span><span class="token punctuation" style="color:#393A34">]</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">ora</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05, minGSSize = 10, maxGSSize = 500&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">gsea</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;pvalueCutoff = 0.05, minGSSize = 10, maxGSSize = 500&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><span class="token key atrule" style="color:#00a4db">annotationforge</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">version</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;0.1.0&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">author</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;tucca-rna-seq &lt;benjamin.bromberg@tufts.edu&gt;&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token string" style="color:#e3116c">&quot;useSynonyms = TRUE&quot;</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token comment" style="color:#999988;font-style:italic"># This section configures local OrgDb package building for organisms</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token comment" style="color:#999988;font-style:italic"># not available in Bioconductor. The workflow automatically detects</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token comment" style="color:#999988;font-style:italic"># if your species needs a local build and creates the package using</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token comment" style="color:#999988;font-style:italic"># AnnotationForge::makeOrgPackage().</span><br></span></code></pre></div></div>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="advanced-configuration-options">Advanced Configuration Options<a href="#advanced-configuration-options" class="hash-link" aria-label="Direct link to Advanced Configuration Options" title="Direct link to Advanced Configuration Options">​</a></h3>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="orgdb-package-management">OrgDb Package Management<a href="#orgdb-package-management" class="hash-link" aria-label="Direct link to OrgDb Package Management" title="Direct link to OrgDb Package Management">​</a></h4>
<p>The workflow automatically handles organism-specific annotation packages:</p>
<ul>
<li><strong>Bioconductor packages</strong>: Automatically installed for supported organisms
(e.g., <code>org.Hs.eg.db</code> for human)</li>
<li><strong>Local builds</strong>: For unsupported organisms, automatically builds OrgDb
packages using AnnotationForge</li>
<li><strong>Species detection</strong>: Automatically determines the correct package based on
your <code>config.yaml</code> species setting</li>
</ul>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="gsea-parameters">GSEA Parameters<a href="#gsea-parameters" class="hash-link" aria-label="Direct link to GSEA Parameters" title="Direct link to GSEA Parameters">​</a></h4>
<div class="language-yaml codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-yaml codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token key atrule" style="color:#00a4db">gsea</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><span class="token key atrule" style="color:#00a4db">gseGO</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">&gt;</span><span class="token punctuation" style="color:#393A34">-</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      pvalueCutoff = 0.05</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      minGSSize = 10</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"> </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      maxGSSize = 500</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      eps = 1e</span><span class="token punctuation" style="color:#393A34">-</span><span class="token number" style="color:#36acaa">10</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      seed = 42</span><br></span></code></pre></div></div>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="ora-parameters">ORA Parameters<a href="#ora-parameters" class="hash-link" aria-label="Direct link to ORA Parameters" title="Direct link to ORA Parameters">​</a></h4>
<div class="language-yaml codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-yaml codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token key atrule" style="color:#00a4db">ora</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><span class="token key atrule" style="color:#00a4db">enrichGO</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">&gt;</span><span class="token punctuation" style="color:#393A34">-</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      pvalueCutoff = 0.05</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      minGSSize = 10</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      maxGSSize = 500</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      qvalueCutoff = 0.2</span><br></span></code></pre></div></div>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="spia-parameters">SPIA Parameters<a href="#spia-parameters" class="hash-link" aria-label="Direct link to SPIA Parameters" title="Direct link to SPIA Parameters">​</a></h4>
<div class="language-yaml codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-yaml codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token key atrule" style="color:#00a4db">spia</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><span class="token key atrule" style="color:#00a4db">extra</span><span class="token punctuation" style="color:#393A34">:</span><span class="token plain"> </span><span class="token punctuation" style="color:#393A34">&gt;</span><span class="token punctuation" style="color:#393A34">-</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    beta = NULL</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    verbose = TRUE</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    plots = FALSE</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    nB = 2000</span><span class="token punctuation" style="color:#393A34">,</span><span class="token plain"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    nPerm = 1000</span><br></span></code></pre></div></div>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="output-files-and-organization">Output Files and Organization<a href="#output-files-and-organization" class="hash-link" aria-label="Direct link to Output Files and Organization" title="Direct link to Output Files and Organization">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="directory-structure">Directory Structure<a href="#directory-structure" class="hash-link" aria-label="Direct link to Directory Structure" title="Direct link to Directory Structure">​</a></h3>
<p>The workflow generates <strong>RDS files only</strong> - no plots, visualizations, or
interactive tools are automatically created:</p>
<div class="language-text codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-text codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">resources/enrichment/</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">├── {analysis_name}/</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">│   └── {contrast_name}/</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">│       ├── gsea_results.RDS          # Gene Set Enrichment Analysis results</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">│       ├── ora_results.RDS           # Over-Representation Analysis results</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">│       ├── spia_results.RDS          # SPIA results (raw)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">│       └── spia_results_readable.RDS # SPIA results (with gene symbols)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">├── spia_data/                        # SPIA data directory (if enabled)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">├── local_orgdb_build/                # Local OrgDb packages (if built)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">└── enrichment_params.RDS             # Enrichment parameters</span><br></span></code></pre></div></div>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="file-formats">File Formats<a href="#file-formats" class="hash-link" aria-label="Direct link to File Formats" title="Direct link to File Formats">​</a></h3>
<ul>
<li><strong>RDS files</strong>: R objects containing complete analysis results</li>
<li><strong>No automatic plots</strong>: The workflow generates data objects, not
visualizations</li>
<li><strong>No PDF/PNG files</strong>: These must be created manually using the R objects</li>
<li><strong>No HTML reports</strong>: Results are stored as R data structures</li>
</ul>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="what-the-workflow-provides">What the Workflow Provides<a href="#what-the-workflow-provides" class="hash-link" aria-label="Direct link to What the Workflow Provides" title="Direct link to What the Workflow Provides">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="1-data-objects-for-analysis">1. <strong>Data Objects for Analysis</strong><a href="#1-data-objects-for-analysis" class="hash-link" aria-label="Direct link to 1-data-objects-for-analysis" title="Direct link to 1-data-objects-for-analysis">​</a></h3>
<p>The workflow generates RDS files containing:</p>
<ul>
<li><strong>GSEA results</strong>: <code>gseaResult</code> objects from clusterProfiler</li>
<li><strong>ORA results</strong>: <code>enrichResult</code> objects from clusterProfiler</li>
<li><strong>SPIA results</strong>: Data frames with pathway impact analysis</li>
<li><strong>MSigDB results</strong>: Gene set enrichment for specified collections</li>
<li><strong>Harmonizome results</strong>: Tissue-specific gene set analysis</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="2-integrated-interactive-analysis-tools">2. <strong>Integrated Interactive Analysis Tools</strong><a href="#2-integrated-interactive-analysis-tools" class="hash-link" aria-label="Direct link to 2-integrated-interactive-analysis-tools" title="Direct link to 2-integrated-interactive-analysis-tools">​</a></h3>
<p>The workflow includes RMarkdown playgrounds that integrate directly with the
outputs:</p>
<ul>
<li><strong>GeneTonic Playground</strong>: Interactive enrichment analysis and visualization</li>
<li><strong>PCA Explorer Playground</strong>: Interactive principal component analysis</li>
<li><strong>Custom Analysis Playground</strong>: Framework for extending analysis</li>
</ul>
<p>These tools automatically load workflow outputs and provide interactive
interfaces through Shiny applications, enabling you to:</p>
<ul>
<li><strong>Create interactive plots</strong> that can be customized and downloaded</li>
<li><strong>Generate publication-ready figures</strong> directly from the apps</li>
<li><strong>Export visualizations</strong> in multiple formats (PNG, PDF, SVG)</li>
<li><strong>Explore data dynamically</strong> before creating final figures</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="3-no-automatic-static-visualizations">3. <strong>No Automatic Static Visualizations</strong><a href="#3-no-automatic-static-visualizations" class="hash-link" aria-label="Direct link to 3-no-automatic-static-visualizations" title="Direct link to 3-no-automatic-static-visualizations">​</a></h3>
<p>The workflow <strong>does not</strong> automatically create:</p>
<ul>
<li>❌ Static plots or bar charts</li>
<li>❌ PDF figures or publication-ready images</li>
<li>❌ Batch-generated visualizations</li>
</ul>
<div class="theme-admonition theme-admonition-info admonition_xJq3 alert alert--info"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 14 16"><path fill-rule="evenodd" d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg></span>Interactive Visualization Through Shiny Apps</div><div class="admonitionContent_BuS1"><p>While the workflow doesn&#x27;t generate static image files, it creates
comprehensive data files (RDS/TSV) that power interactive Shiny applications.
These apps allow you to:</p><ul>
<li><strong>Explore data interactively</strong> through dynamic plots and tables</li>
<li><strong>Download visualizations</strong> as PNG, PDF, or other image formats</li>
<li><strong>Create publication-ready figures</strong> directly from the interactive interfaces</li>
<li><strong>Customize plots</strong> with interactive controls before downloading</li>
</ul></div></div>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="4-analysis-workflow">4. <strong>Analysis Workflow</strong><a href="#4-analysis-workflow" class="hash-link" aria-label="Direct link to 4-analysis-workflow" title="Direct link to 4-analysis-workflow">​</a></h3>
<p>The complete analysis process:</p>
<ol>
<li><strong>Workflow execution</strong>: Generates data objects and HTML reports</li>
<li><strong>Interactive analysis</strong>: Use integrated RMarkdown playgrounds</li>
<li><strong>Custom visualization</strong>: Create publication-ready figures manually</li>
<li><strong>Export results</strong>: Share data files and custom visualizations</li>
</ol>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="5-automatic-orgdb-management">5. <strong>Automatic OrgDb Management</strong><a href="#5-automatic-orgdb-management" class="hash-link" aria-label="Direct link to 5-automatic-orgdb-management" title="Direct link to 5-automatic-orgdb-management">​</a></h3>
<p>The workflow automatically handles organism-specific annotation packages:</p>
<ul>
<li><strong>Detection</strong>: Identifies whether your species has a Bioconductor package</li>
<li><strong>Installation</strong>: Installs Bioconductor packages automatically</li>
<li><strong>Local building</strong>: Creates custom OrgDb packages for unsupported organisms</li>
<li><strong>Integration</strong>: Seamlessly provides the correct annotation data for
enrichment analysis</li>
</ul>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="interpreting-results">Interpreting Results<a href="#interpreting-results" class="hash-link" aria-label="Direct link to Interpreting Results" title="Direct link to Interpreting Results">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="gsea-results">GSEA Results<a href="#gsea-results" class="hash-link" aria-label="Direct link to GSEA Results" title="Direct link to GSEA Results">​</a></h3>
<p><strong>Key Metrics:</strong></p>
<ul>
<li><strong>Enrichment Score (ES)</strong>: Measure of gene set enrichment</li>
<li><strong>Normalized Enrichment Score (NES)</strong>: Normalized ES for comparison</li>
<li><strong>P-value</strong>: Statistical significance</li>
<li><strong>FDR q-value</strong>: False discovery rate adjusted p-value</li>
</ul>
<p><strong>Interpretation:</strong></p>
<ul>
<li><strong>Positive ES</strong>: Genes upregulated in the gene set</li>
<li><strong>Negative ES</strong>: Genes downregulated in the gene set</li>
<li><strong>|NES| &gt; 1.5</strong>: Strong enrichment</li>
<li><strong>FDR &lt; 0.25</strong>: Statistically significant</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="ora-results">ORA Results<a href="#ora-results" class="hash-link" aria-label="Direct link to ORA Results" title="Direct link to ORA Results">​</a></h3>
<p><strong>Key Metrics:</strong></p>
<ul>
<li><strong>Gene Ratio</strong>: Proportion of genes in the gene set</li>
<li><strong>P-value</strong>: Statistical significance</li>
<li><strong>Adjusted P-value</strong>: Multiple testing correction</li>
<li><strong>Count</strong>: Number of genes in the intersection</li>
</ul>
<p><strong>Interpretation:</strong></p>
<ul>
<li><strong>Gene Ratio &gt; 0.1</strong>: High proportion of genes</li>
<li><strong>Adjusted P-value &lt; 0.05</strong>: Statistically significant</li>
<li><strong>Count &gt; 5</strong>: Sufficient genes for reliable results</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="spia-results">SPIA Results<a href="#spia-results" class="hash-link" aria-label="Direct link to SPIA Results" title="Direct link to SPIA Results">​</a></h3>
<p><strong>Key Metrics:</strong></p>
<ul>
<li><strong>PERT</strong>: Perturbation statistic</li>
<li><strong>PGFdr</strong>: False discovery rate adjusted p-value</li>
<li><strong>Status</strong>: Pathway activation/inhibition</li>
<li><strong>tA</strong>: Total perturbation accumulation</li>
</ul>
<p><strong>Interpretation:</strong></p>
<ul>
<li><strong>Status = &quot;Activated&quot;</strong>: Pathway is upregulated</li>
<li><strong>Status = &quot;Inhibited&quot;</strong>: Pathway is downregulated</li>
<li><strong>PGFdr &lt; 0.05</strong>: Statistically significant pathway</li>
</ul>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="working-with-results">Working with Results<a href="#working-with-results" class="hash-link" aria-label="Direct link to Working with Results" title="Direct link to Working with Results">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="loading-results-in-r">Loading Results in R<a href="#loading-results-in-r" class="hash-link" aria-label="Direct link to Loading Results in R" title="Direct link to Loading Results in R">​</a></h3>
<div class="language-r codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-r codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Load enrichment results</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(tidyverse)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># GSEA results</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gsea_results &lt;- readRDS(&quot;resources/enrichment/analysis_name/contrast_name/gsea_results.RDS&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># ORA results</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ora_results &lt;- readRDS(&quot;resources/enrichment/analysis_name/contrast_name/ora_results.RDS&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># SPIA results</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">spia_results &lt;- readRDS(&quot;resources/enrichment/analysis_name/contrast_name/spia_results_readable.RDS&quot;)</span><br></span></code></pre></div></div>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="interactive-analysis-with-integrated-tools">Interactive Analysis with Integrated Tools<a href="#interactive-analysis-with-integrated-tools" class="hash-link" aria-label="Direct link to Interactive Analysis with Integrated Tools" title="Direct link to Interactive Analysis with Integrated Tools">​</a></h3>
<p>The workflow includes RMarkdown playgrounds that integrate directly with the
generated outputs:</p>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="genetonic-playground">GeneTonic Playground<a href="#genetonic-playground" class="hash-link" aria-label="Direct link to GeneTonic Playground" title="Direct link to GeneTonic Playground">​</a></h4>
<p>The <code>analysis/GeneTonic_playground.Rmd</code> provides interactive enrichment
analysis:</p>
<div class="language-r codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-r codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Load workflow outputs</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dds_wald &lt;- readRDS(&quot;resources/deseq2/analysis_name/dds.RDS&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">res_de &lt;- readRDS(&quot;resources/deseq2/analysis_name/contrast_name/wald.RDS&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">res_enrich &lt;- readRDS(&quot;resources/enrichment/analysis_name/contrast_name/ora_results.RDS&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Launch interactive GeneTonic app</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gtl &lt;- GeneTonic::GeneTonicList(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  dds = dds_wald,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  res_de = res_de,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  res_enrich = res_enrich_shaken,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  annotation_obj = annotation_obj</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">GeneTonic::GeneTonic(gtl = gtl, project_id = &quot;My Project&quot;)</span><br></span></code></pre></div></div>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="pca-explorer-playground">PCA Explorer Playground<a href="#pca-explorer-playground" class="hash-link" aria-label="Direct link to PCA Explorer Playground" title="Direct link to PCA Explorer Playground">​</a></h4>
<p>The <code>analysis/pcaExplorer_playground.Rmd</code> provides interactive PCA analysis:</p>
<div class="language-r codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-r codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Load workflow outputs</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dds &lt;- readRDS(&quot;resources/deseq2/analysis_name/dds.RDS&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dst &lt;- readRDS(&quot;resources/deseq2/analysis_name/dst.RDS&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Launch interactive pcaExplorer app</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">pcaExplorer::pcaExplorer(</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  dds = dds,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  dst = dst,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  countmatrix = countmatrix,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  coldata = coldata,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  annotation = annotation_obj</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre></div></div>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="ideal-integration">Ideal Integration<a href="#ideal-integration" class="hash-link" aria-label="Direct link to Ideal Integration" title="Direct link to Ideal Integration">​</a></h4>
<p>The workflow outputs are also compatible with the <code>ideal</code> package for
interactive differential expression analysis:</p>
<div class="language-r codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-r codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Load workflow outputs</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dds &lt;- readRDS(&quot;resources/deseq2/analysis_name/dds.RDS&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">res &lt;- readRDS(&quot;resources/deseq2/analysis_name/contrast_name/wald.RDS&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Launch interactive ideal app</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ideal::ideal(dds = dds, res = res)</span><br></span></code></pre></div></div>
<div class="theme-admonition theme-admonition-info admonition_xJq3 alert alert--info"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 14 16"><path fill-rule="evenodd" d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg></span>Planned Feature</div><div class="admonitionContent_BuS1"><p>A dedicated <code>ideal</code> playground RMarkdown file is planned for future releases to
provide the same integrated experience as the existing GeneTonic and
pcaExplorer playgrounds.</p></div></div>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="manual-visualization-creation">Manual Visualization Creation<a href="#manual-visualization-creation" class="hash-link" aria-label="Direct link to Manual Visualization Creation" title="Direct link to Manual Visualization Creation">​</a></h3>
<p>You can also create custom visualizations manually using the data objects:</p>
<div class="language-r codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_QJqH"><pre tabindex="0" class="prism-code language-r codeBlock_bY9V thin-scrollbar" style="color:#393A34;background-color:#f6f8fa"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Load required libraries</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(clusterProfiler)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(enrichplot)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Create dot plot for GO results</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">if (!is.null(ora_results$GO)) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  dotplot(ora_results$GO, showCategory = 20)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Create enrichment map for GSEA results</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">if (!is.null(gsea_results$GO)) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  emapplot(pairwise_termsim(gsea_results$GO))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Create network plot</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">if (!is.null(ora_results$KEGG)) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  cnetplot(ora_results$KEGG, showCategory = 10)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre></div></div>
<div class="theme-admonition theme-admonition-tip admonition_xJq3 alert alert--success"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Interactive vs. Manual Visualization</div><div class="admonitionContent_BuS1"><p><strong>Interactive Tools (Recommended)</strong>: Use GeneTonic and pcaExplorer for
exploration and downloading publication-ready figures directly from the apps.</p><p><strong>Manual Creation</strong>: Use the code examples above when you need custom plots
or want to integrate visualizations into R scripts or reports.</p></div></div>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="best-practices">Best Practices<a href="#best-practices" class="hash-link" aria-label="Direct link to Best Practices" title="Direct link to Best Practices">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="1-parameter-selection">1. <strong>Parameter Selection</strong><a href="#1-parameter-selection" class="hash-link" aria-label="Direct link to 1-parameter-selection" title="Direct link to 1-parameter-selection">​</a></h3>
<ul>
<li><strong>P-value cutoff</strong>: 0.05 for discovery, 0.01 for validation</li>
<li><strong>Gene set size</strong>: 10-500 genes for reliable results</li>
<li><strong>Multiple testing</strong>: Use FDR correction for multiple comparisons</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="2-quality-control">2. <strong>Quality Control</strong><a href="#2-quality-control" class="hash-link" aria-label="Direct link to 2-quality-control" title="Direct link to 2-quality-control">​</a></h3>
<ul>
<li><strong>Background genes</strong>: Use all expressed genes as background</li>
<li><strong>Gene set overlap</strong>: Avoid highly overlapping gene sets</li>
<li><strong>Statistical power</strong>: Ensure sufficient sample size</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="3-interpretation">3. <strong>Interpretation</strong><a href="#3-interpretation" class="hash-link" aria-label="Direct link to 3-interpretation" title="Direct link to 3-interpretation">​</a></h3>
<ul>
<li><strong>Biological relevance</strong>: Focus on biologically meaningful results</li>
<li><strong>Pathway consistency</strong>: Look for consistent patterns across analyses</li>
<li><strong>Validation</strong>: Confirm key findings with independent methods</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="4-visualization">4. <strong>Visualization</strong><a href="#4-visualization" class="hash-link" aria-label="Direct link to 4-visualization" title="Direct link to 4-visualization">​</a></h3>
<ul>
<li><strong>Interactive exploration</strong>: Start with GeneTonic and pcaExplorer for data
exploration</li>
<li><strong>Download from apps</strong>: Export publication-ready figures directly from the
Shiny interfaces</li>
<li><strong>Manual creation</strong>: Use clusterProfiler functions for custom plots when
needed</li>
</ul>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="troubleshooting">Troubleshooting<a href="#troubleshooting" class="hash-link" aria-label="Direct link to Troubleshooting" title="Direct link to Troubleshooting">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="common-issues">Common Issues<a href="#common-issues" class="hash-link" aria-label="Direct link to Common Issues" title="Direct link to Common Issues">​</a></h3>
<table><thead><tr><th>Problem</th><th>Cause</th><th>Solution</th></tr></thead><tbody><tr><td><strong>No significant results</strong></td><td>P-value too strict</td><td>Increase p-value cutoff</td></tr><tr><td><strong>Too many results</strong></td><td>P-value too lenient</td><td>Decrease p-value cutoff</td></tr><tr><td><strong>Empty gene sets</strong></td><td>Background too small</td><td>Use all expressed genes</td></tr><tr><td><strong>SPIA errors</strong></td><td>KEGG API issues</td><td>Check internet connection</td></tr></tbody></table>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="getting-help">Getting Help<a href="#getting-help" class="hash-link" aria-label="Direct link to Getting Help" title="Direct link to Getting Help">​</a></h3>
<p>If you encounter issues:</p>
<ol>
<li><strong>Check configuration</strong>: Verify YAML syntax and parameters</li>
<li><strong>Review logs</strong>: Check Snakemake and R logs</li>
<li><strong>Validate input</strong>: Ensure differential expression results exist</li>
<li><strong>Open issue</strong>: Report problems on <a href="https://github.com/tucca-cellag/tucca-rna-seq/issues" target="_blank" rel="noopener noreferrer">GitHub</a></li>
</ol>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="next-steps">Next Steps<a href="#next-steps" class="hash-link" aria-label="Direct link to Next Steps" title="Direct link to Next Steps">​</a></h2>
<p>After completing enrichment analysis:</p>
<ol>
<li><strong>Load Results</strong>: Use <code>readRDS()</code> to load the generated data objects</li>
<li><strong>Interactive Exploration</strong>: Use GeneTonic and pcaExplorer for data
exploration</li>
<li><strong>Download Visualizations</strong>: Export publication-ready figures from the
Shiny apps</li>
<li><strong>Custom Analysis</strong>: Perform additional statistical analyses or create
custom plots</li>
<li><strong>Biological Interpretation</strong>: Relate findings to your research question</li>
</ol>
<p>For detailed workflow execution, see the
<a href="/tucca-rna-seq/running">Running Guide</a>.</p>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="summary">Summary<a href="#summary" class="hash-link" aria-label="Direct link to Summary" title="Direct link to Summary">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="what-the-workflow-provides-1">What the Workflow Provides<a href="#what-the-workflow-provides-1" class="hash-link" aria-label="Direct link to What the Workflow Provides" title="Direct link to What the Workflow Provides">​</a></h3>
<p>✅ <strong>Comprehensive data objects</strong> for all enrichment analyses<br>
<!-- -->✅ <strong>Statistical results</strong> ready for interpretation<br>
<!-- -->✅ <strong>Gene ID mappings</strong> and annotations<br>
<!-- -->✅ <strong>Flexible configuration</strong> for different analysis types<br>
<!-- -->✅ <strong>Integrated RMarkdown playgrounds</strong> for interactive analysis<br>
<!-- -->✅ <strong>HTML reports</strong> (FastQC, Qualimap, MultiQC)</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="what-the-workflow-does-not-provide">What the Workflow Does NOT Provide<a href="#what-the-workflow-does-not-provide" class="hash-link" aria-label="Direct link to What the Workflow Does NOT Provide" title="Direct link to What the Workflow Does NOT Provide">​</a></h3>
<p>❌ <strong>Automatic static plots or visualizations</strong><br>
<!-- -->❌ <strong>Batch-generated figures</strong><br>
<!-- -->❌ <strong>Publication-ready images</strong></p>
<div class="theme-admonition theme-admonition-info admonition_xJq3 alert alert--info"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 14 16"><path fill-rule="evenodd" d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg></span>But You Can Create Them!</div><div class="admonitionContent_BuS1"><p>The workflow generates the data needed to create visualizations through:</p><ul>
<li><strong>Interactive Shiny apps</strong> (GeneTonic, pcaExplorer) that allow plot download</li>
<li><strong>R data objects</strong> that can be used with plotting libraries</li>
<li><strong>Integrated playgrounds</strong> that provide visualization capabilities</li>
</ul></div></div>
<p><strong>The workflow provides both data generation AND integrated interactive
analysis tools.</strong></p>
<hr>
<p><em>Functional enrichment analysis provides biological context for your gene
expression changes. The workflow generates comprehensive data objects and
integrates with interactive analysis tools like GeneTonic, pcaExplorer, and
ideal, allowing you to explore your results interactively and create custom
visualizations for publication.</em></p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="row margin-top--sm theme-doc-footer-edit-meta-row"><div class="col"></div><div class="col lastUpdated_JAkA"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-09-24T17:19:22.000Z" itemprop="dateModified">Sep 24, 2025</time></b> by <b>benjibromberg</b></span></div></div></footer></article><nav class="docusaurus-mt-lg pagination-nav" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/tucca-rna-seq/running"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Running tucca-rna-seq</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/tucca-rna-seq/tufts-specific/hpc-quick-start"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Tufts HPC Quick Start Guide</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#overview" class="table-of-contents__link toc-highlight">Overview</a></li><li><a href="#analysis-types" class="table-of-contents__link toc-highlight">Analysis Types</a><ul><li><a href="#1-standard-enrichment-analysis-clusterprofiler" class="table-of-contents__link toc-highlight">1. Standard Enrichment Analysis (clusterProfiler)</a></li><li><a href="#2-msigdb-molecular-signatures-database" class="table-of-contents__link toc-highlight">2. MSigDB (Molecular Signatures Database)</a></li><li><a href="#3-spia-signaling-pathway-impact-analysis" class="table-of-contents__link toc-highlight">3. SPIA (Signaling Pathway Impact Analysis)</a></li><li><a href="#4-harmonizome-database" class="table-of-contents__link toc-highlight">4. Harmonizome Database</a></li></ul></li><li><a href="#configuration" class="table-of-contents__link toc-highlight">Configuration</a><ul><li><a href="#basic-configuration" class="table-of-contents__link toc-highlight">Basic Configuration</a></li><li><a href="#advanced-configuration-options" class="table-of-contents__link toc-highlight">Advanced Configuration Options</a></li></ul></li><li><a href="#output-files-and-organization" class="table-of-contents__link toc-highlight">Output Files and Organization</a><ul><li><a href="#directory-structure" class="table-of-contents__link toc-highlight">Directory Structure</a></li><li><a href="#file-formats" class="table-of-contents__link toc-highlight">File Formats</a></li></ul></li><li><a href="#what-the-workflow-provides" class="table-of-contents__link toc-highlight">What the Workflow Provides</a><ul><li><a href="#1-data-objects-for-analysis" class="table-of-contents__link toc-highlight">1. <strong>Data Objects for Analysis</strong></a></li><li><a href="#2-integrated-interactive-analysis-tools" class="table-of-contents__link toc-highlight">2. <strong>Integrated Interactive Analysis Tools</strong></a></li><li><a href="#3-no-automatic-static-visualizations" class="table-of-contents__link toc-highlight">3. <strong>No Automatic Static Visualizations</strong></a></li><li><a href="#4-analysis-workflow" class="table-of-contents__link toc-highlight">4. <strong>Analysis Workflow</strong></a></li><li><a href="#5-automatic-orgdb-management" class="table-of-contents__link toc-highlight">5. <strong>Automatic OrgDb Management</strong></a></li></ul></li><li><a href="#interpreting-results" class="table-of-contents__link toc-highlight">Interpreting Results</a><ul><li><a href="#gsea-results" class="table-of-contents__link toc-highlight">GSEA Results</a></li><li><a href="#ora-results" class="table-of-contents__link toc-highlight">ORA Results</a></li><li><a href="#spia-results" class="table-of-contents__link toc-highlight">SPIA Results</a></li></ul></li><li><a href="#working-with-results" class="table-of-contents__link toc-highlight">Working with Results</a><ul><li><a href="#loading-results-in-r" class="table-of-contents__link toc-highlight">Loading Results in R</a></li><li><a href="#interactive-analysis-with-integrated-tools" class="table-of-contents__link toc-highlight">Interactive Analysis with Integrated Tools</a></li><li><a href="#manual-visualization-creation" class="table-of-contents__link toc-highlight">Manual Visualization Creation</a></li></ul></li><li><a href="#best-practices" class="table-of-contents__link toc-highlight">Best Practices</a><ul><li><a href="#1-parameter-selection" class="table-of-contents__link toc-highlight">1. <strong>Parameter Selection</strong></a></li><li><a href="#2-quality-control" class="table-of-contents__link toc-highlight">2. <strong>Quality Control</strong></a></li><li><a href="#3-interpretation" class="table-of-contents__link toc-highlight">3. <strong>Interpretation</strong></a></li><li><a href="#4-visualization" class="table-of-contents__link toc-highlight">4. <strong>Visualization</strong></a></li></ul></li><li><a href="#troubleshooting" class="table-of-contents__link toc-highlight">Troubleshooting</a><ul><li><a href="#common-issues" class="table-of-contents__link toc-highlight">Common Issues</a></li><li><a href="#getting-help" class="table-of-contents__link toc-highlight">Getting Help</a></li></ul></li><li><a href="#next-steps" class="table-of-contents__link toc-highlight">Next Steps</a></li><li><a href="#summary" class="table-of-contents__link toc-highlight">Summary</a><ul><li><a href="#what-the-workflow-provides-1" class="table-of-contents__link toc-highlight">What the Workflow Provides</a></li><li><a href="#what-the-workflow-does-not-provide" class="table-of-contents__link toc-highlight">What the Workflow Does NOT Provide</a></li></ul></li></ul></div></div></div></div></main></div></div></div><footer class="theme-layout-footer footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/">Home</a></li><li class="footer__item"><a class="footer__link-item" href="/tucca-rna-seq/introduction">tucca-rna-seq</a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Connect With Us</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://cellularagriculture.tufts.edu/" target="_blank" rel="noopener noreferrer" class="footer__link-item">Tufts University Center for Cellular Agriculture<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li><li class="footer__item"><a href="https://www.linkedin.com/company/tufts-cell-ag/" target="_blank" rel="noopener noreferrer" class="footer__link-item">TUCCA on LinkedIn<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li><li class="footer__item"><a href="https://twitter.com/tuftscellag" target="_blank" rel="noopener noreferrer" class="footer__link-item">TUCCA on X<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li><li class="footer__item"><a href="https://www.youtube.com/channel/UC29F8uqsu_K7aRxOgjfG_HQ" target="_blank" rel="noopener noreferrer" class="footer__link-item">TUCCA on YouTube<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/tucca-cellag" target="_blank" rel="noopener noreferrer" class="footer__link-item">TUCCA&#x27;s GitHub<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Tufts University
      Center for Cellular Agriculture. Built with Docusaurus.</div></div></div></footer></div>
</body>
</html>