<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper plugin-docs plugin-id-default docs-version-current docs-doc-page docs-doc-id-tucca-rna-seq/install-and-config/deployment" data-has-hydrated="false">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v3.8.1">
<title data-rh="true">Workflow Deployment Options | TUCCA Bioinformatics Docs</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:image" content="https://tucca-cellag.github.io/img/Center for Cellular Agriculture_blk+bl_vert.png"><meta data-rh="true" name="twitter:image" content="https://tucca-cellag.github.io/img/Center for Cellular Agriculture_blk+bl_vert.png"><meta data-rh="true" property="og:url" content="https://tucca-cellag.github.io/tucca-rna-seq/install-and-config/deployment"><meta data-rh="true" property="og:locale" content="en"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Workflow Deployment Options | TUCCA Bioinformatics Docs"><meta data-rh="true" name="description" content="Recommended Deployment Strategy"><meta data-rh="true" property="og:description" content="Recommended Deployment Strategy"><link data-rh="true" rel="canonical" href="https://tucca-cellag.github.io/tucca-rna-seq/install-and-config/deployment"><link data-rh="true" rel="alternate" href="https://tucca-cellag.github.io/tucca-rna-seq/install-and-config/deployment" hreflang="en"><link data-rh="true" rel="alternate" href="https://tucca-cellag.github.io/tucca-rna-seq/install-and-config/deployment" hreflang="x-default"><script data-rh="true" type="application/ld+json">{"@context":"https://schema.org","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Workflow Deployment Options","item":"https://tucca-cellag.github.io/tucca-rna-seq/install-and-config/deployment"}]}</script><link rel="stylesheet" href="/assets/css/styles.2d474d0a.css">
<script src="/assets/js/runtime~main.23394ea9.js" defer="defer"></script>
<script src="/assets/js/main.29038db3.js" defer="defer"></script>
</head>
<body class="navigation-with-keyboard">
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;"><defs>
<symbol id="theme-svg-external-link" viewBox="0 0 24 24"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"/></symbol>
</defs></svg>
<script>!function(){var t="light";var e=function(){try{return new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}}()||function(){try{return window.localStorage.getItem("theme")}catch(t){}}();document.documentElement.setAttribute("data-theme",e||t),document.documentElement.setAttribute("data-theme-choice",e||t)}(),function(){try{const c=new URLSearchParams(window.location.search).entries();for(var[t,e]of c)if(t.startsWith("docusaurus-data-")){var a=t.replace("docusaurus-data-","data-");document.documentElement.setAttribute(a,e)}}catch(t){}}()</script><div id="__docusaurus"><div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="theme-layout-navbar navbar navbar--fixed-top"><div class="navbar__inner"><div class="theme-layout-navbar-left navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/"><b class="navbar__title text--truncate">TUCCA Bioinformatics Docs</b></a><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/tucca-rna-seq/introduction">tucca-rna-seq</a><a class="navbar__item navbar__link" href="/helpful-resources/learn-about-cell-ag">Helpful Resources</a><a class="navbar__item navbar__link" href="/publications/ast1">Publications</a><a class="navbar__item navbar__link" href="/our-team">Our Team</a></div><div class="theme-layout-navbar-right navbar__items navbar__items--right"><a href="https://github.com/tucca-cellag" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">Our GitHub Organization<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="system mode" aria-label="Switch between dark and light mode (currently system mode)"><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" aria-hidden="true" class="toggleIcon_g3eP systemToggleIcon_QzmC"><path fill="currentColor" d="m12 21c4.971 0 9-4.029 9-9s-4.029-9-9-9-9 4.029-9 9 4.029 9 9 9zm4.95-13.95c1.313 1.313 2.05 3.093 2.05 4.95s-0.738 3.637-2.05 4.95c-1.313 1.313-3.093 2.05-4.95 2.05v-14c1.857 0 3.637 0.737 4.95 2.05z"></path></svg></button></div><div class="navbarSearchContainer_Bca1"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="theme-layout-main main-wrapper mainWrapper_z2l0"><div class="docsWrapper_hBAB"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docRoot_UBD9"><aside class="theme-doc-sidebar-container docSidebarContainer_YfHR"><div class="sidebarViewport_aRkj"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/tucca-rna-seq/introduction">Introduction</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a href="https://github.com/tucca-cellag/tucca-rna-seq" target="_blank" rel="noopener noreferrer" class="menu__link menuExternalLink_NmtK">GitHub Repository<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" role="button" aria-expanded="false" href="/tucca-rna-seq/data-collection/data-collection-why">Data Collection 101 for RNA-Seq</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" role="button" aria-expanded="true" href="/tucca-rna-seq/install-and-config/snakemake-primer">Installation and Configuration of tucca-rna-seq</a></div><ul class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/tucca-rna-seq/install-and-config/snakemake-primer">Snakemake Primer</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/tucca-rna-seq/install-and-config/deployment">Workflow Deployment Options</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/tucca-rna-seq/install-and-config/installation">Installing tucca-rna-seq</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/tucca-rna-seq/install-and-config/configuration">Configuring tucca-rna-seq for your Analysis</a></li></ul></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/tucca-rna-seq/running">Running tucca-rna-seq</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/tucca-rna-seq/enrichment-analysis">Functional Enrichment Analysis</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" href="/tucca-rna-seq/tufts-specific/hpc-quick-start">For Tufts Users</a><button aria-label="Expand sidebar category &#x27;For Tufts Users&#x27;" aria-expanded="false" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/tucca-rna-seq/citing-the-workflow">Citing the Workflow</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/tucca-rna-seq/help">Getting Help</a></li></ul></nav></div></div></aside><main class="docMainContainer_TBSr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">Installation and Configuration of tucca-rna-seq</span></li><li class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link">Workflow Deployment Options</span></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><header><h1>Workflow Deployment Options</h1></header>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="recommended-deployment-strategy">Recommended Deployment Strategy<a href="#recommended-deployment-strategy" class="hash-link" aria-label="Direct link to Recommended Deployment Strategy" title="Direct link to Recommended Deployment Strategy">​</a></h2>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="layered-environments-with-singularityapptainer--conda">Layered Environments with Singularity/Apptainer + Conda<a href="#layered-environments-with-singularityapptainer--conda" class="hash-link" aria-label="Direct link to Layered Environments with Singularity/Apptainer + Conda" title="Direct link to Layered Environments with Singularity/Apptainer + Conda">​</a></h3>
<p>The most robust and reproducible way to run this workflow is by combining
containers and package management in a layered approach, as described in the
<strong><a href="https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html#ad-hoc-combination-of-conda-package-management-with-containers" target="_blank" rel="noopener noreferrer">official Snakemake documentation</a></strong>. This is the method we&#x27;ve
implemented and recommend.</p>
<p>At first, it might sound redundant: &quot;Why use a package manager <em>inside</em> a
container?&quot; The answer lies in achieving the ultimate balance of
reproducibility, flexibility, and efficiency.</p>
<p>Let&#x27;s continue with our kitchen and shopping list analogy:</p>
<ol>
<li>
<p><strong><code>Singularity/Apptainer</code> is the Portable Kitchen:</strong> It provides the entire
standardized kitchen environment: the operating system, the plumbing, the
wiring, and the core <code>conda</code> appliance. When you run the workflow with
<code>--use-singularity</code> or <code>--use-apptainer</code>, you are ensuring every step of the
analysis happens inside this identical, portable kitchen, no matter if you&#x27;re
on your laptop or an HPC cluster.</p>
</li>
<li>
<p><strong><code>--use-conda</code> is the Automated Shopping Trip:</strong> For every single step in
the main recipe (a Snakemake rule), <code>--use-conda</code> performs a specialized,
just-in-time shopping trip. It reads a very specific shopping list (the rule&#x27;s
<code>environment.yaml</code> file) that details the exact &quot;brand&quot; and &quot;version&quot; of every
ingredient (software like <code>fastqc=0.11.9</code>) needed for <em>only that step</em>. It then
uses the <code>conda</code> tool inside the kitchen to instantly procure these items and
place them in a clean, isolated pantry for that specific task.</p>
</li>
</ol>
<p>This two-layer system is the gold standard for several reasons:</p>
<ul>
<li><strong>Absolute Reproducibility:</strong> You control the outer environment (the
kitchen&#x27;s OS via Singularity/Apptainer) and the inner environment (the specific
ingredients for each step via Conda). This eliminates almost every variable
that could cause results to differ between machines.</li>
<li><strong>Unmatched Flexibility:</strong> Need one brand of flour (<code>samtools v1.9</code>) for the
cake and a different brand (<code>samtools v1.15</code>) for the bread? No problem. Each
recipe step gets its own shopping trip and its own isolated pantry, so
ingredients never get mixed up.</li>
<li><strong>Efficiency:</strong> The base kitchen (Apptainer container) stays small and
simple; it only needs the basic <code>conda</code> appliance. You don&#x27;t need to stock the
kitchen with every possible ingredient from the start. The workflow performs
these small, fast shopping trips on-the-fly, only when an ingredient is needed.</li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="how-it-works-in-practice">How It Works in Practice<a href="#how-it-works-in-practice" class="hash-link" aria-label="Direct link to How It Works in Practice" title="Direct link to How It Works in Practice">​</a></h3>
<p>When you execute the command <code>snakemake all --use-singularity --use-conda</code>, to
run a configured workflow, a precise sequence of events unfolds:</p>
<ol>
<li>
<p><strong>DAG Calculation:</strong> First, Snakemake reads the entire <code>Snakefile</code> (the
recipe). It compares the outputs you&#x27;ve asked for with the files that already
exist, and it builds a full dependency graph of all jobs that need to run. This
is the <strong>Directed Acyclic Graph (DAG)</strong>, which serves as the master plan for
the entire analysis.</p>
</li>
<li>
<p><strong>Environment Creation:</strong> Next, <em>before a single
computational job is run</em>, Snakemake identifies all the unique software
environments (<code>environment.yaml</code> files) required for the jobs in the DAG. It
then creates all of these Conda environments. This is like doing all the
grocery shopping for every course of a multi-course meal before you even start
preheating the oven.</p>
</li>
<li>
<p><strong>Job Execution:</strong> With the plan (DAG) made and all the pantries stocked
(Conda environments created), Snakemake begins executing jobs.</p>
<ul>
<li>It starts with the jobs at the beginning of the DAG, which have no
unfinished dependencies.</li>
<li>For each job, it uses <strong>Singularity/Apptainer</strong> to enter the standardized
&quot;kitchen.&quot;</li>
<li>Inside, it activates the specific, pre-built <strong>Conda</strong> environment
(&quot;pantry&quot;) for that job.</li>
<li>It runs the job&#x27;s script in this doubly-isolated environment.</li>
<li>As each job finishes, Snakemake checks the DAG to see which downstream
jobs are now unlocked and ready to run, continuing until the final output
is generated.</li>
</ul>
</li>
</ol>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="helpful-resources-docs--guides">Helpful Resources (Docs / Guides)<a href="#helpful-resources-docs--guides" class="hash-link" aria-label="Direct link to Helpful Resources (Docs / Guides)" title="Direct link to Helpful Resources (Docs / Guides)">​</a></h2>
<div class="container center"><div class="row" style="margin-bottom:20px"><div class="col" style="display:flex"><div class="card item shadow--md" style="margin-top:20px;flex-grow:1"><div class="card__header"><h3 class="anchor anchorWithStickyNavbar_LWe7" id="apptainer--singularity">Apptainer / Singularity<a href="#apptainer--singularity" class="hash-link" aria-label="Direct link to Apptainer / Singularity" title="Direct link to Apptainer / Singularity">​</a></h3></div><div class="card__body"><p></p><p>The container runtime that allows you to run the workflow in a
self-contained environment. We use it to run a Docker container image.</p><p></p><a href="https://apptainer.org/docs/user/main/" target="_blank" rel="noopener noreferrer"><button class="button button--primary button--block" role="button" aria-disabled="false">Apptainer User Guide</button></a><a href="https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html#running-jobs-in-containers" target="_blank" rel="noopener noreferrer"><button class="button button--secondary button--block" style="margin-top:10px" role="button" aria-disabled="false">Snakemake &amp; Containers</button></a></div><div class="card__footer"><p style="font-size:0.8em;font-style:italic"></p><p>A quick note: Apptainer is the new name for the open-source container
technology formerly known as Singularity. For the purposes of this
workflow, the tools are functionally identical, and we may use their
names interchangeably.</p><p></p></div></div></div><div class="col" style="display:flex"><div class="card item shadow--md" style="margin-top:20px;flex-grow:1"><div class="card__header"><h3 class="anchor anchorWithStickyNavbar_LWe7" id="conda">Conda<a href="#conda" class="hash-link" aria-label="Direct link to Conda" title="Direct link to Conda">​</a></h3></div><div class="card__body"><p></p><p>The package manager that operates <strong>inside the container</strong>
to manage the specific software packages required by the Snakemake
rules.</p><p></p><a href="https://docs.conda.io/projects/conda/en/latest/user-guide/index.html" target="_blank" rel="noopener noreferrer"><button class="button button--primary button--block" role="button" aria-disabled="false">Conda User Guide</button></a><a href="https://hub.docker.com/r/continuumio/miniconda3" target="_blank" rel="noopener noreferrer"><button class="button button--secondary button--block" style="margin-top:10px" role="button" aria-disabled="false">Default container: continuumio/miniconda3</button></a></div><div class="card__footer"><p style="font-size:0.8em;font-style:italic"></p><p>Note: There are several Conda distributions, such as Anaconda
(includes many scientific packages by default), Miniconda (a minimal
installer), and Miniforge/Mambaforge (uses the community-driven
conda-forge channel). Our workflow&#x27;s container uses Miniconda.</p><p></p></div></div></div></div></div>
<div class="theme-admonition theme-admonition-warning admonition_xJq3 alert alert--warning"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 16 16"><path fill-rule="evenodd" d="M8.893 1.5c-.183-.31-.52-.5-.887-.5s-.703.19-.886.5L.138 13.499a.98.98 0 0 0 0 1.001c.193.31.53.501.886.501h13.964c.367 0 .704-.19.877-.5a1.03 1.03 0 0 0 .01-1.002L8.893 1.5zm.133 11.497H6.987v-2.003h2.039v2.003zm0-3.004H6.987V5.987h2.039v4.006z"></path></svg></span>A Note on Installation</div><div class="admonitionContent_BuS1"><p>Installing Apptainer/Singularity requires administrator (root) privileges,
which can make it difficult to set up on a personal machine.</p><p>However, because it is a standard tool in scientific computing, most HPC
clusters already have it installed. We recommend checking your HPC&#x27;s
documentation or contacting its support team to confirm before you start.</p></div></div>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="choosing-an-execution-platform">Choosing an Execution Platform<a href="#choosing-an-execution-platform" class="hash-link" aria-label="Direct link to Choosing an Execution Platform" title="Direct link to Choosing an Execution Platform">​</a></h2>
<p>In addition to the software deployment strategy (using Conda, Singularity,
or both), you also need to decide on the hardware or platform where the workflow
will run. Snakemake&#x27;s plugin-based architecture allows it to adapt to a wide
variety of execution and storage backends.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="execution-backends">Execution Backends<a href="#execution-backends" class="hash-link" aria-label="Direct link to Execution Backends" title="Direct link to Execution Backends">​</a></h3>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="high-performance-computing-hpc-clusters">High-Performance Computing (HPC) Clusters<a href="#high-performance-computing-hpc-clusters" class="hash-link" aria-label="Direct link to High-Performance Computing (HPC) Clusters" title="Direct link to High-Performance Computing (HPC) Clusters">​</a></h4>
<p>A <strong>High-Performance Compute (HPC)</strong> cluster is a network of powerful servers
designed for complex data analysis. Many researchers use HPC clusters to run
bioinformatics workflows because they can process large datasets much faster
than a personal computer. The workflow can be adapted to different HPC
schedulers by creating custom profiles and using the appropriate executor
plugin:</p>
<ul>
<li><strong>Slurm</strong>: <code>snakemake-executor-plugin-slurm</code> (used in this workflow&#x27;s
pre-configured profiles)</li>
<li><strong>LSF</strong>: <code>snakemake-executor-plugin-lsf</code></li>
<li><strong>PBS/Torque</strong>: <code>snakemake-executor-plugin-pbs</code></li>
<li><strong>SGE</strong>: <code>snakemake-executor-plugin-sge</code></li>
<li>And many others, which you can explore in the
<a href="https://snakemake.github.io/snakemake-plugin-catalog/#executor-plugins" target="_blank" rel="noopener noreferrer">Snakemake Executor Plugin Catalog</a>.</li>
</ul>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="local-execution">Local Execution<a href="#local-execution" class="hash-link" aria-label="Direct link to Local Execution" title="Direct link to Local Execution">​</a></h4>
<p>For small datasets or testing, you can run the workflow on a local machine
(e.g., a personal laptop or workstation). In this case, Snakemake does not
require a special executor plugin and will run jobs directly on your computer.</p>
<div class="theme-admonition theme-admonition-warning admonition_xJq3 alert alert--warning"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 16 16"><path fill-rule="evenodd" d="M8.893 1.5c-.183-.31-.52-.5-.887-.5s-.703.19-.886.5L.138 13.499a.98.98 0 0 0 0 1.001c.193.31.53.501.886.501h13.964c.367 0 .704-.19.877-.5a1.03 1.03 0 0 0 .01-1.002L8.893 1.5zm.133 11.497H6.987v-2.003h2.039v2.003zm0-3.004H6.987V5.987h2.039v4.006z"></path></svg></span>Local Execution Limitations</div><div class="admonitionContent_BuS1"><p>Local execution is suitable for small genomes and limited datasets. For
production analyses, an HPC cluster or cloud environment is recommended.</p></div></div>
<h4 class="anchor anchorWithStickyNavbar_LWe7" id="cloud-and-container-orchestration">Cloud and Container Orchestration<a href="#cloud-and-container-orchestration" class="hash-link" aria-label="Direct link to Cloud and Container Orchestration" title="Direct link to Cloud and Container Orchestration">​</a></h4>
<p>The workflow also supports various cloud and container platforms through
dedicated executor plugins:</p>
<ul>
<li><strong>Kubernetes</strong>: <code>snakemake-executor-plugin-kubernetes</code></li>
<li><strong>Google Cloud Batch</strong>: <code>snakemake-executor-plugin-googlebatch</code></li>
<li><strong>AWS Batch</strong>: <code>snakemake-executor-plugin-aws-batch</code></li>
</ul>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="continuous-integration-ci-with-github-actions">Continuous Integration (CI) with GitHub Actions<a href="#continuous-integration-ci-with-github-actions" class="hash-link" aria-label="Direct link to Continuous Integration (CI) with GitHub Actions" title="Direct link to Continuous Integration (CI) with GitHub Actions">​</a></h3>
<p>This repository comes with a pre-configured GitHub Actions workflow (<code>.github/workflows/main.yml</code>)
that automatically tests the workflow&#x27;s integrity on every push and pull
request. This ensures that the code is always in a working state.</p>
<div class="theme-admonition theme-admonition-danger admonition_xJq3 alert alert--danger"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M5.05.31c.81 2.17.41 3.38-.52 4.31C3.55 5.67 1.98 6.45.9 7.98c-1.45 2.05-1.7 6.53 3.53 7.7-2.2-1.16-2.67-4.52-.3-6.61-.61 2.03.53 3.33 1.94 2.86 1.39-.47 2.3.53 2.27 1.67-.02.78-.31 1.44-1.13 1.81 3.42-.59 4.78-3.42 4.78-5.56 0-2.84-2.53-3.22-1.25-5.61-1.52.13-2.03 1.13-1.89 2.75.09 1.08-1.02 1.8-1.86 1.33-.67-.41-.66-1.19-.06-1.78C8.18 5.31 8.68 2.45 5.05.32L5.03.3l.02.01z"></path></svg></span>For Testing Only</div><div class="admonitionContent_BuS1"><p>The included GitHub Actions workflow runs on free, public runners, which have
limited computational resources and disk space. <strong>It is designed for testing
with small datasets only and cannot be used to run a real-world analysis.</strong></p><p>For large-scale production runs, you can repurpose the
<code>.github/workflows/main.yml</code> file by configuring it to use <strong>self-hosted
runners</strong> with sufficient resources. This typically involves changing the
<code>runs-on</code> key in the workflow file to target your self-hosted runner group.</p></div></div>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="storage-backends">Storage Backends<a href="#storage-backends" class="hash-link" aria-label="Direct link to Storage Backends" title="Direct link to Storage Backends">​</a></h3>
<p>For workflows that run in the cloud or across different physical locations,
Snakemake can use <strong>storage plugins</strong> to seamlessly access data from various
remote storage backends. This means your <code>Snakefile</code> can reference remote files
(e.g., in an S3 bucket) as if they were on your local filesystem.</p>
<p>Some of the available storage plugins include:</p>
<ul>
<li><strong>Amazon S3</strong>: <code>snakemake-storage-plugin-s3</code></li>
<li><strong>Google Cloud Storage</strong>: <code>snakemake-storage-plugin-gcs</code></li>
<li><strong>Microsoft Azure Blob Storage</strong>: <code>snakemake-storage-plugin-azure</code></li>
</ul>
<p>You can explore all available storage backends in the
<a href="https://snakemake.github.io/snakemake-plugin-catalog/#storage-plugins" target="_blank" rel="noopener noreferrer">Snakemake Storage Plugin Catalog</a>.</p>
<hr>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="working-on-a-high-performance-compute-hpc-cluster">Working on a High-Performance Compute (HPC) Cluster<a href="#working-on-a-high-performance-compute-hpc-cluster" class="hash-link" aria-label="Direct link to Working on a High-Performance Compute (HPC) Cluster" title="Direct link to Working on a High-Performance Compute (HPC) Cluster">​</a></h2>
<p>You typically connect to an HPC cluster using <strong>SSH (Secure Shell)</strong>, which
provides a secure command-line interface to the remote servers.</p>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="hpc-learning-materials">HPC Learning Materials<a href="#hpc-learning-materials" class="hash-link" aria-label="Direct link to HPC Learning Materials" title="Direct link to HPC Learning Materials">​</a></h3>
<div class="theme-admonition theme-admonition-info admonition_xJq3 alert alert--info"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 14 16"><path fill-rule="evenodd" d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg></span>New to HPC?</div><div class="admonitionContent_BuS1"><p>If you are not familiar with high-performance computing, these resources are a
great place to start.</p><div class="tabs-container tabList__CuJ"><ul role="tablist" aria-orientation="horizontal" class="tabs"><li role="tab" tabindex="0" aria-selected="true" class="tabs__item tabItem_LNqP tabs__item--active">HPC Carpentry Lessons</li><li role="tab" tabindex="-1" aria-selected="false" class="tabs__item tabItem_LNqP">Tufts HPC Home Page</li><li role="tab" tabindex="-1" aria-selected="false" class="tabs__item tabItem_LNqP">Tufts HPC User Guide</li></ul><div class="margin-top--md"><div role="tabpanel" class="tabItem_Ymn6"><p></p><p>HPC Carpentry teaches foundational skills for high-performance computing.</p><p></p><a href="https://www.hpc-carpentry.org/community-lessons/" target="_blank" rel="noopener noreferrer"><button class="button button--primary button--block" role="button" aria-disabled="false">Visit HPC Carpentry</button></a></div><div role="tabpanel" class="tabItem_Ymn6" hidden=""><p></p><p>Tufts Technology Service&#x27;s (TTS) home page for the Tufts University HPC
cluster.</p><p></p><a href="https://it.tufts.edu/high-performance-computing" target="_blank" rel="noopener noreferrer"><button class="button button--primary button--block" role="button" aria-disabled="false">Go to HPC Home Page</button></a></div><div role="tabpanel" class="tabItem_Ymn6" hidden=""><p></p><p>Technical documentation for the Tufts HPC cluster, provided by TTS
Research Technology.</p><p></p><a href="https://rtguides.it.tufts.edu/hpc/" target="_blank" rel="noopener noreferrer"><button class="button button--primary button--block" role="button" aria-disabled="false">Read the HPC User Guide</button></a></div></div></div></div></div>
<div class="theme-admonition theme-admonition-tip admonition_xJq3 alert alert--success"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>Tufts HPC Materials</div><div class="admonitionContent_BuS1"><p>If you are Tufts staff/student/faculty, please refer to our Tufts-specific
guides for deployment via the Tufts HPC cluster:</p><a href="/tucca-rna-seq/tufts-specific/hpc-quick-start"><button class="button button--primary button--block" role="button" aria-disabled="false">Tufts HPC Quick Start Guide</button></a><a href="/tucca-rna-seq/tufts-specific/hpc-best-practices"><button class="button button--secondary button--block" style="margin-top:10px" role="button" aria-disabled="false">Tufts HPC Best Practices</button></a></div></div>
<h2 class="anchor anchorWithStickyNavbar_LWe7" id="recommended-tool-visual-studio-code-vscode">Recommended Tool: Visual Studio Code (VSCode)<a href="#recommended-tool-visual-studio-code-vscode" class="hash-link" aria-label="Direct link to Recommended Tool: Visual Studio Code (VSCode)" title="Direct link to Recommended Tool: Visual Studio Code (VSCode)">​</a></h2>
<p>We highly recommend using an integrated development environment (IDE) like
<strong>VSCode</strong> to interact with this workflow, especially when working on an HPC
cluster.</p>
<p><strong>VSCode</strong> is a free, powerful code editor that can connect to remote servers
via SSH. This allows you to edit files, run commands, and manage your workflow
on the HPC cluster directly from a user-friendly interface on your local
machine. It offers several key advantages for scientific workflows:</p>
<ul>
<li><strong>Remote Development</strong>: VSCode can seamlessly connect to remote servers,
allowing you to edit files, run code, and manage projects directly on the
HPC without leaving your local environment.</li>
<li><strong>Integrated Terminal</strong>: Access the HPC terminal within VSCode, enabling
efficient command-line operations alongside your coding activities.</li>
<li><strong>Extensions and Customization</strong>: Enhance your development experience with
extensions tailored for specific programming languages, debugging tools, and
workflow optimizations.</li>
</ul>
<div class="theme-admonition theme-admonition-tip admonition_xJq3 alert alert--success"><div class="admonitionHeading_Gvgb"><span class="admonitionIcon_Rf37"><svg viewBox="0 0 12 16"><path fill-rule="evenodd" d="M6.5 0C3.48 0 1 2.19 1 5c0 .92.55 2.25 1 3 1.34 2.25 1.78 2.78 2 4v1h5v-1c.22-1.22.66-1.75 2-4 .45-.75 1-2.08 1-3 0-2.81-2.48-5-5.5-5zm3.64 7.48c-.25.44-.47.8-.67 1.11-.86 1.41-1.25 2.06-1.45 3.23-.02.05-.02.11-.02.17H5c0-.06 0-.13-.02-.17-.2-1.17-.59-1.83-1.45-3.23-.2-.31-.42-.67-.67-1.11C2.44 6.78 2 5.65 2 5c0-2.2 2.02-4 4.5-4 1.22 0 2.36.42 3.22 1.19C10.55 2.94 11 3.94 11 5c0 .66-.44 1.78-.86 2.48zM4 14h5c-.23 1.14-1.3 2-2.5 2s-2.27-.86-2.5-2z"></path></svg></span>tip</div><div class="admonitionContent_BuS1"><p>If you are Tufts staff/student/faculty, please refer our documentation for
setting up access to the Tufts HPC cluster via VSCode:</p><a href="/tucca-rna-seq/tufts-specific/tufts-hpc-in-vscode"><button class="button button--primary" role="button" aria-disabled="false">Learn to Set Up Tufts HPC Access in VSCode</button></a></div></div>
<h3 class="anchor anchorWithStickyNavbar_LWe7" id="learning-resources-for-vscode">Learning Resources for VSCode<a href="#learning-resources-for-vscode" class="hash-link" aria-label="Direct link to Learning Resources for VSCode" title="Direct link to Learning Resources for VSCode">​</a></h3>
<div class="container center"><div class="row"><div class="col"><div class="card item shadow--md" style="margin-top:20px"><div class="card__header"><h3>Official Documentation</h3></div><div class="card__body"><p></p><p>The official docs are the best place to start, covering everything
from basic setup to advanced features.</p><p></p><a href="https://code.visualstudio.com/docs" target="_blank" rel="noopener noreferrer"><button class="button button--primary button--block" role="button" aria-disabled="false">VS Code Docs</button></a><a href="https://code.visualstudio.com/docs/getstarted/getting-started" target="_blank" rel="noopener noreferrer"><button class="button button--secondary button--block" style="margin-top:10px" role="button" aria-disabled="false">Getting Started Guide</button></a><a href="https://code.visualstudio.com/docs/getstarted/introvideos" target="_blank" rel="noopener noreferrer"><button class="button button--info button--block" style="margin-top:10px" role="button" aria-disabled="false">Introductory Videos</button></a></div></div></div><div class="col"><div class="card item shadow--md" style="margin-top:20px"><div class="card__header"><h3>VS Code in 100 Seconds</h3></div><div class="card__body"><div style="width:100%;height:100%"></div></div><div class="card__footer"><p></p><p>A quick overview of VS Code&#x27;s core features, including the command
palette, integrated terminal, and extensions.</p><p></p></div></div></div><div class="col"><div class="card item shadow--md" style="margin-top:20px"><div class="card__header"><h3>25 VS Code Productivity Tips</h3></div><div class="card__body"><div style="width:100%;height:100%"></div></div><div class="card__footer"><p></p><p>Improve your workflow with tips on keyboard shortcuts, multi-line
editing, and useful extensions like GitLens and Quokka.</p><p></p></div></div></div></div></div>
<div class="container center"><div class="row"><div class="col"><div class="card item shadow--md" style="margin-top:20px"><div class="card__header"><h3 class="anchor anchorWithStickyNavbar_LWe7" id="download-and-install-vscode">Download and Install VSCode<a href="#download-and-install-vscode" class="hash-link" aria-label="Direct link to Download and Install VSCode" title="Direct link to Download and Install VSCode">​</a></h3></div><div class="card__body"><p></p><p>Visit the official Visual Studio Code website to download the
installer for your operating system.</p><p></p><a href="https://code.visualstudio.com/download" target="_blank" rel="noopener noreferrer"><button class="button button--primary button--block" role="button" aria-disabled="false">VSCode Download Page</button></a><div style="margin-top:20px;text-align:left"><h4>Installation Instructions:</h4><details class="details_lb9f alert alert--info details_b_Ee" data-collapsed="true"><summary>Windows</summary><div><div class="collapsibleContent_i85q"><ul><li>Run the downloaded <code>.exe</code> installer.</li><li>Follow the installation prompts, accepting the license agreement and selecting desired installation options (e.g., adding VSCode to your PATH).</li></ul></div></div></details><details class="details_lb9f alert alert--info details_b_Ee" data-collapsed="true"><summary>macOS</summary><div><div class="collapsibleContent_i85q"><ul><li>Open the downloaded <code>.dmg</code> file.</li><li>Drag and drop the VSCode application into the Applications
folder.</li></ul></div></div></details><details class="details_lb9f alert alert--info details_b_Ee" data-collapsed="true"><summary>Linux</summary><div><div class="collapsibleContent_i85q"><p></p><p>Follow the instructions for your distribution on the VSCode
website.</p><p></p><a href="https://code.visualstudio.com/docs/setup/linux" target="_blank" rel="noopener noreferrer"><button class="button button--secondary button--block" role="button" aria-disabled="false">Linux Setup Instructions</button></a></div></div></details></div></div><div class="card__footer"><p></p><p>After installation, open VSCode from your application launcher or by
running <code>code</code> in your terminal (if added to PATH).</p><p></p></div></div></div></div></div></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="row margin-top--sm theme-doc-footer-edit-meta-row"><div class="col"></div><div class="col lastUpdated_JAkA"><span class="theme-last-updated">Last updated<!-- --> on <b><time datetime="2025-09-24T17:19:22.000Z" itemprop="dateModified">Sep 24, 2025</time></b> by <b>benjibromberg</b></span></div></div></footer></article><nav class="docusaurus-mt-lg pagination-nav" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/tucca-rna-seq/install-and-config/snakemake-primer"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Snakemake Primer</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/tucca-rna-seq/install-and-config/installation"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Installing tucca-rna-seq</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#recommended-deployment-strategy" class="table-of-contents__link toc-highlight">Recommended Deployment Strategy</a><ul><li><a href="#layered-environments-with-singularityapptainer--conda" class="table-of-contents__link toc-highlight">Layered Environments with Singularity/Apptainer + Conda</a></li><li><a href="#how-it-works-in-practice" class="table-of-contents__link toc-highlight">How It Works in Practice</a></li></ul></li><li><a href="#helpful-resources-docs--guides" class="table-of-contents__link toc-highlight">Helpful Resources (Docs / Guides)</a><ul><li><a href="#apptainer--singularity" class="table-of-contents__link toc-highlight">Apptainer / Singularity</a></li><li><a href="#conda" class="table-of-contents__link toc-highlight">Conda</a></li></ul></li><li><a href="#choosing-an-execution-platform" class="table-of-contents__link toc-highlight">Choosing an Execution Platform</a><ul><li><a href="#execution-backends" class="table-of-contents__link toc-highlight">Execution Backends</a></li><li><a href="#continuous-integration-ci-with-github-actions" class="table-of-contents__link toc-highlight">Continuous Integration (CI) with GitHub Actions</a></li><li><a href="#storage-backends" class="table-of-contents__link toc-highlight">Storage Backends</a></li></ul></li><li><a href="#working-on-a-high-performance-compute-hpc-cluster" class="table-of-contents__link toc-highlight">Working on a High-Performance Compute (HPC) Cluster</a><ul><li><a href="#hpc-learning-materials" class="table-of-contents__link toc-highlight">HPC Learning Materials</a></li></ul></li><li><a href="#recommended-tool-visual-studio-code-vscode" class="table-of-contents__link toc-highlight">Recommended Tool: Visual Studio Code (VSCode)</a><ul><li><a href="#learning-resources-for-vscode" class="table-of-contents__link toc-highlight">Learning Resources for VSCode</a></li><li><a href="#download-and-install-vscode" class="table-of-contents__link toc-highlight">Download and Install VSCode</a></li></ul></li></ul></div></div></div></div></main></div></div></div><footer class="theme-layout-footer footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/">Home</a></li><li class="footer__item"><a class="footer__link-item" href="/tucca-rna-seq/introduction">tucca-rna-seq</a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">Connect With Us</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://cellularagriculture.tufts.edu/" target="_blank" rel="noopener noreferrer" class="footer__link-item">Tufts University Center for Cellular Agriculture<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li><li class="footer__item"><a href="https://www.linkedin.com/company/tufts-cell-ag/" target="_blank" rel="noopener noreferrer" class="footer__link-item">TUCCA on LinkedIn<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li><li class="footer__item"><a href="https://twitter.com/tuftscellag" target="_blank" rel="noopener noreferrer" class="footer__link-item">TUCCA on X<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li><li class="footer__item"><a href="https://www.youtube.com/channel/UC29F8uqsu_K7aRxOgjfG_HQ" target="_blank" rel="noopener noreferrer" class="footer__link-item">TUCCA on YouTube<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li></ul></div><div class="theme-layout-footer-column col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/tucca-cellag" target="_blank" rel="noopener noreferrer" class="footer__link-item">TUCCA&#x27;s GitHub<svg width="13.5" height="13.5" aria-hidden="true" class="iconExternalLink_nPIU"><use href="#theme-svg-external-link"></use></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2025 Tufts University
      Center for Cellular Agriculture. Built with Docusaurus.</div></div></div></footer></div>
</body>
</html>